### DISCUSSION TITLE
Why is there a math function on Energy Usage / Charge?

### DISCUSSION THREAD

**Discussion by youzer-name** [jasonacox/Powerwall-Dashboard#355] [General]
@jasonacox  I just had a third Powerwall installed today and noticed that in the Energy Usage panel, the Charge value dipped below zero during the period when the installers had the Powerwalls off but hadn't yet killed power to the gateway (and Powerwall Dashboard system). This seems to be because the panel is set up to use the percentage field with "math(/ 0.95 - 5/0.95)" as a modifier and the gateway was reporting zero charge level during that time.

When the percentage = 100 it returns 100, but for every value below that it returns a lower result than the value in the database. It returns zero at 5%, and negative values for anything below 5%. What is the purpose of using that equation on the values returned from the database?

**Comment by BJReplay** [comment]
I might have suggested that function, but even if I didn't, I can explain it.

The percentage charge displayed in the Tesla app is different to the percentage charge returned by the API.

That's because the system reserves 5% of the capacity for re-starting your solar inverters if the Powerwall discharges to zero, and you're off grid.  It uses this to power up an islanded grid and restart a solar inverter to re-charge off-grid.

The app appears to internally use this function (either in the app, or in the back end) to adjust the values it gets from the Powerwall when it returns the charge it displays in the app.

The idea behind that function is to align the two values so that what you see in the dashboard matches (as closely as reasonably possible) to what you'd see in the App if you were to open the App at the same time.

You'd rarely notice any discrepancy, and even more rarely notice any negative value, because it requires the powerwalls to discharged to zero, but, as you know, it can happen.

**Reply by jasonacox** [reply]
@BJReplay is correct. We could explore a way to filter out the negative data when the system is doing an upgrade/offline.

**Reply by mcbirse** [reply]
There is always the option to filter this using a Telegraf Starlark Processor script... I am doing this so do not see the negative charge percentages when that occurs (often during firmware updates).

There was an example of the script I'm using in this discussion: #112
